[RDY] Allow resizing the terminal when setting &columns or &lines, or using :winsize, also in startup.

fmoralesc opened the pull request:
Closes #2863

fmoralesc committed:
Allow resizing the terminal window

Neither setting the 'columns' and 'lines' options nor using the `:winsize` command
resized the terminal window, which caused display glitches.

Re: #2863

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -350,6 +352,13 @@ static void tui_resize(UI *ui, int width, int height)
+  if (term && (STARTS_WITH(term, "xterm") || STARTS_WITH(term, "rxvt"))) {
+      // try to resize the terminal window
+      char r[16]; // enough for 9999x9999
+      sprintf(r, "\x1b[8;%d;%dt", height, width);
```
What happens if we send this to a terminal that _doesn't_ support resizing? Is it just ignored? I'm wondering if we can skip the `STARTS_WITH` checks, which will turn into a mess as we keep adding checks all over the place...

fmoralesc replied:
I'm not sure, actually, I just wanted to be safe... 

I just tried it with the linux terminal, and it simply ignored the control sequence, so I think removing the check is safe. 

I think libvterm doesn't set $TERM by itself, but I'm not sure.

justinmk commented:
It's not clear to me from https://github.com/neovim/neovim/issues/2863 how to test this. Would it be possible to write a test, or at least mention steps to reproduce in the commit message?

fmoralesc commented:
Oh, sure, my bad.

Anyway, to test this, try different variations of the `winsize` cmd, e.g.: `:winsize 20 40`, `:winsize 100 40`, etc.

fmoralesc commented:
@justinmk It seems like this can't be tested using the expect functions in the functional tests suite. The screen data neovim provides *is* correct, so this:
~~~ lua
  describe(':set co= lines= :winsize', function()
    describe(':set co= lines=', function()
      it('resize the UI', function()
	execute('set co=20 lines=10')
	screen:expect([[
	^                    |
	~                   |
	~                   |
	~                   |
	~                   |
	~                   |
	~                   |
	~                   |
	~                   |
	                    |
	]])
      end)
    end)
    describe(':winsize', function()
      it('resizes the UI', function()
	execute('winsize 15 8')
	screen:expect([[
	^               |
	~              |
	~              |
	~              |
	~              |
	~              |
	~              |
	               |
	]])
      end)
    end)
  end)
~~~
works as expected both with this PR and without it. However, the side effects of theses operations in a real terminal are not accounted for. For example, I get this without these patches:

![captura de pantalla de 2015-07-18 19-29-26](https://cloud.githubusercontent.com/assets/221465/8763831/a3f40b30-2d86-11e5-991b-4640a47c365a.png)

As you can see, the text grid is resized, but the terminal window doesn't.

Testing this, there seems to be a secondary issue: the `&columns` and `&lines` options are not respected at startup, so having
~~~ vim
set co=100
set lines=30
~~~
in the vimrc doesn't have any effect, this workaround has to be used:
~~~ vim
au VimEnter * set co=100 lines=30
~~~
It seems like this is because of the behavior of the `update_size()` function.

fmoralesc committed:
Make the TUI respect the 'co' and 'lines' options on startup

`nvim --cmd "set co=... lines="` didn't work as expected, and forced to
set those options on VimEnter or afterwards.

fmoralesc commented:
I included a check so it will resize correctly on startup. However, the new logic will get confused when the passed values for `columns` or `lines` are the defaults (80 and 42) (for example, with `nvim --cmd "set co=80 lines=24`), since it can't tell if these values are the defaults (in which case it should override them) or are set by the user (in which case it should use them). 

Anyway, I think the behavior is much improved.

fmoralesc commented:
This also improves the behaviour of nvim when sourcing sessions, and might be a fix for https://github.com/neovim/neovim/issues/2097 and https://github.com/neovim/neovim/issues/1482

mhinz commented:
This doesn't seem to work for iTerm2. I'm looking into it.

fmoralesc commented:
@mhinz Do you want me to leave a note in the documentation about that?

mhinz commented:
I'm not sure, since this applies to all commandline programs. Maybe this should go into the "resize" section of some kind of FAQ?

fmoralesc commented:
> runtime/doc/nvim-faq.txt, anyone?

:+1:

mhinz commented:
I think this is RDY, eh?

fmoralesc commented:
@mhinz I think so too, I was waiting for a second review just in case. I must admit, I've been very tempted to merged this, this bug has been driving me mad for a while ;)

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -641,12 +645,21 @@ static void update_size(UI *ui)
+
+  // 1 - if starting and the 'columns' and 'lines' options are not
+  // the defaults, use their values
+  if (starting && (Columns != 80 || Rows != 24)) {
```
can't we avoid hard-coding the values here?

justinmk replied:
The defaults of 80 and 24 are again used on lines 679-680, so we can at least define them at the top of this function.

fmoralesc replied:
Let me see what I can do. It would be best if we defined sth that would get used here and in `option.c`.

justinmk replied:
Ok. By the way, I just noticed that `starting` is not a boolean.

```c
/* Values for "starting" */
#define NO_SCREEN       2       /* no screen updating yet */
#define NO_BUFFERS      1       /* not all buffers loaded yet */
/*			0	   not starting anymore */
```

fmoralesc replied:
I only care that it isn't 0 here, do you want me to make that explicit?

justinmk replied:
Yes, that's how other cases are.

fmoralesc committed:
options,tui: don't hardcode default terminal size

also, include some checks.

fmoralesc commented:
OK. Made the suggested changes (thanks for taking a look btw, @justinmk), so it should be good now.

Pyrohh commented on src/nvim/tui/tui.c:
```diff
@@ -641,12 +646,21 @@ static void update_size(UI *ui)
+  // 1 - if starting and the 'columns' and 'lines' options are not
+  // the defaults, use their values
+  if (starting != 0 && (Columns != DFLT_COLS || Rows != DFLT_LINES)) {
+    width = (int)Columns;
```
Maybe add an assert here? Just as a sanity check.

fmoralesc replied:
Not sure about this. What check do you think is worth it here?

Pyrohh replied:
`Columns` is of type `long`, so isn't guaranteed to fit into an `int`. I recall it being discussed earlier that `Columns` and `Rows` should really be of type `int`, but that might be rather invasive of a refactoring.

fmoralesc replied:
Ah, sure. :+1: I was thinking of something else.

fmoralesc replied:
Done.

Pyrohh commented:
Besides the tiny nit, LGTM.

justinmk commented on src/nvim/option_defs.h:
```diff
@@ -59,6 +59,9 @@
 #define EOL_MAC         2       /* CR */
 
+#define DFLT_COLS       80      // default value for 'columns'
+#define DFLT_LINES      24      // default value for 'lines'
```
What about the `Columns` and `Rows` declarations in `globals.h`?

fmoralesc replied:
Forgot about those! I'm away from my laptop, so If you think this is mergeable I would kindly ask you to make that change. Or we can wait ;)

justinmk replied:
No worries, I am away as well though.

fmoralesc replied:
OK, done.

justinmk commented:
Merged with some changes:

- renamed DFLT_LINES to DFLT_ROWS
- changed inline macro comments to `/**/` instead of `//`

fmoralesc commented:
Thanks @justinmk

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -350,6 +351,10 @@ static void tui_resize(UI *ui, int width, int height)
   data->row = data->col = 0;
+  // try to resize the terminal window
+  char r[16];  // enough for 9999x9999
+  snprintf(r, sizeof(r), "\x1b[8;%d;%dt", height, width);
```
@fmoralesc If I'm reading http://invisible-island.net/xterm/ctlseqs/ctlseqs.html correctly it seems we should use `4` instead of `8`. Do you recall why you choose `8`?

```
CSI Ps ; Ps ; Ps t
          Window manipulation (from dtterm, as well as extensions).
          ...
            Ps = 4  ;  height ;  width -> Resize the xterm window to
          given height and width in pixels.  Omitted parameters reuse
          the current height or width.  Zero parameters use the dis-
          play's height or width.
            ...
            Ps = 8  ;  height ;  width -> Resize the text area to given
          height and width in characters.  Omitted parameters reuse the
          current height or width.  Zero parameters use the display's
          height or width.
```

ZyX-I replied:
Pixels does not look much useful: most of time UI uses character width/height. Depends though how it is supposed to be used: users may prefer pixels (though not all: I do not know how wide some of my terminals are in pixels, yet I normally know how wide they are in characters). Can’t find code where UI resize function receives exactly characters.

justinmk replied:
Thanks @ZyX-I. I missed the word "pixels", I was focused on "window" vs "text area" and I took the latter to mean something else.
